### ISSUE TITLE
NaN gradients not properly masked

### ISSUE THREAD

**Issue by wyndwarrior** [pytorch/pytorch#17455]
## 🐛 Bug

After performing some computation that results in a NaN gradient, no amount of reassignment will result in a non-NaN gradient.

## To Reproduce

```import torch
import numpy as np

x = torch.from_numpy(np.array([0, 100])).float()
x.requires_grad = True
f = torch.log((torch.exp(x) + 1))

f[1] = 0
f[:] = 0

clean_f = torch.zeros(100)
clean_f[0] = f[0]

print(torch.autograd.grad(f[1], x, retain_graph=True))
print(torch.autograd.grad(f[0], x, retain_graph=True))
print(torch.autograd.grad(clean_f[1], x, retain_graph=True))
print(f)
print(torch.__version__)
```

Produces:

```
(tensor([0., nan]),)
(tensor([0., nan]),)
(tensor([0., nan]),)
tensor([0., 0.], grad_fn=<CopySlices>)
1.0.1.post2
```

## Expected behavior

```
(tensor([0., 0.]),)
(tensor([0., 0.]),)
(tensor([0., 0.]),)
tensor([0., 0.], grad_fn=<CopySlices>)
1.0.1.post2
```

## Environment
PyTorch version: 1.0.1.post2
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 16.04.4 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609
CMake version: version 2.8.12.2

Python version: 3.6
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti

Nvidia driver version: 410.79
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.5.1.10
/usr/lib/x86_64-linux-gnu/libcudnn.so.6.0.21

Versions of relevant libraries:
[pip3] numpy==1.12.0
[conda] blas                      1.0                         mkl  
[conda] cuda100                   1.0                           0    pytorch
[conda] cuda80                    1.0                  h205658b_0    pytorch
[conda] mkl                       2019.1                      144  
[conda] mkl_fft                   1.0.6            py36hd81dba3_0  
[conda] mkl_random                1.0.2            py36hd81dba3_0  
[conda] pytorch                   1.0.1           py3.6_cuda10.0.130_cudnn7.4.2_2    pytorch
[conda] torchvision               0.2.1                      py_2    pytorch

**Comment by t-vi** [comment]
This is the same idea as #15506 and #16317. Short of keeping a mask around for all backward steps, it's pretty hard to solve as it doesn't fit floating point arithmetic.

**Comment by soumith** [comment]
See Alban's comments here: https://github.com/pytorch/pytorch/issues/15506#issuecomment-450544895 on why this happens and our reasoning / stance on it.
